File: agent/crates/fleet_client/src/enrollment.py
from typing import Any

from fleet-client.types import (
    RegisterRequest,
    EnrollmentResult,
)


class AgentEnrollment:
    @staticmethod
    async def enroll(stub: Any, request: RegisterRequest) -> EnrollmentResult:
        """
        Register the agent with the Fleet server.

        Args:
            stub: FleetService gRPC stub.
            request: RegisterRequest.

        Returns:
            EnrollmentResult
        """

        response = await stub.RegisterAgent(request)

        return EnrollmentResult(
            node_id=response.node_id,
            token=response.token,
            config=response.config,
        )

File: agent/crates/fleet_client/src/enrollment.rs
use crate::types::{EnrollmentResult, RegisterRequest, RegisterResponse};
use anyhow::Result;
use tonic::transport::Channel;
use edr_sdk::codec::JsonCodec;

pub struct AgentEnrollment;

impl AgentEnrollment {
    pub async fn enroll(channel: Channel, request: RegisterRequest) -> Result<EnrollmentResult> {
        tracing::info!("Enrolling agent: {:?}", request.hostname);

        let mut client = tonic::client::Grpc::new(channel);
        let path = http::uri::PathAndQuery::from_static("/edr.fleet.FleetService/RegisterAgent");
        let tonic_req = tonic::Request::new(request);
        
        let response = client
            .unary(
                tonic_req,
                path,
                JsonCodec::<RegisterRequest, RegisterResponse>::default()
            )
            .await?;

        let res = response.into_inner();

        Ok(EnrollmentResult {
            node_id: res.node_id,
            token: res.token,
            config: res.config,
        })
    }
}
